### ISSUE TITLE
Excessive CPU time in ivrclientcore_get_generic_interface

### ISSUE THREAD

**Issue by ryao** [ValveSoftware/SteamVR-for-Linux#148]
#### How To Collect SteamVR System Information:
1. Click the 'SteamVR beta' menu in the upper left of the SteamVR Status window
2. Click 'Create System Report'
3. Click 'Copy to Clipboard'
4. Browse to https://gist.github.com/
5. Click in the text entry box (by the 1)
6. Paste clipboard contents
7. Click 'Create Public Gist'
8. Add URL of your new Gist below

#### Your system information
* Steam client version (build number or date): 153939410
* Distribution (e.g. Ubuntu): Gentoo
* Graphics driver version (run nvidia-settings): 396.54.09
* Gist for SteamVR System Information: https://gist.github.com/ryao/787519c68533865934bcde0e3cf9cb8c
* Opted into Steam client beta?: Yes
* Opted into SteamVR beta?: No
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:
The Lab.exe is CPU bound on a single core, causing GPU pipeline stalls with Proton 3.16-3 beta. This is not a regression. It has been around since Proton first debuted, but various issues kept me from feeling comfortable filing a bug report due to difficulty doing decent triage.

Today, I was able to profile it with perf in a way that reveals the issue. Here is a flame graph demonstrating it:

![Flame graph](https://cdn.discordapp.com/attachments/490220625764548609/502288668698869770/lab.svg)

For those unfamiliar with flame graphs, the Y axis is stack frames of stacks. The X axis is the percentage of samples a given stack was seen. You can read more about it here:

https://github.com/brendangregg/FlameGraph

Anyway, this shows excessive CPU time in ivrclientcore_get_generic_interface, or more specifically, in IPC::CreateMutexWithNamespace. 72% of CPU time is being spent creating mutexes. I have no idea why it is doing that. The source code is not public, so I cannot triage/debug further.

#### Steps for reproducing this issue:
0. Install Linux
1. Install Steam
2. Install SteamVR
3. Install The Lab
4. Try playing it.
5. While it is running, profile using `perf record -ag -F 997 -- sleep 10`.
6. Then you can use something like `perf script | FlameGraph/stackcollapse-perf.pl | FlameGraph/flamegraph.pl --title='The Lab on Proton 3.16-3: 10 seconds idle in Longbow sampled at 997Hz' > /tmp/lab.svg` to get a flame graph showing the issue.

You will need this patch in order for the stackcollapse-perf.pl script to work:

https://github.com/brendangregg/FlameGraph/pull/185

**Comment by h1z1** [comment]
See #143

**Comment by ryao** [comment]
@h1z1 I think that is a separate issue. I tried making `.local/share/Steam/steamapps/common/SteamVR/bin/linux64/chmod_trace_marker` into a setuid program to see if it makes a difference and it just isn't being executed. The flame graph showing __GI___kill taking most of our CPU time implies that this issue has nothing to do with `chmod_trace_marker`.

**Comment by ryao** [comment]
I did some more profiling. These are the syscalls being executed by TheLab.exe over a 10 second interval:

```
SYSCALL                   COUNT
kill                   10961289
readv                    165884
poll                     126740
write                     52679
getpid                    25323
clock_gettime              5377
futex                      4482
sched_yield                3319
select                     3058
openat                     1932
```
It is doing ~1.1 million kill syscalls per second, which is absurd. ^_^;;

**Comment by ryao** [comment]
Just a FYI, the call to kill is being done in `IPC::CreateMutex`. The call to `kill@@GLIBC_2.2.5` goes through the Global Offset Table. When control moves to the kill function, it immediately executes these two instructions:

```
   362c0:       b8 3e 00 00 00          mov    $0x3e,%eax
   362c5:       0f 05                   syscall
```
The consequence is that the stack unwinder does not recognize the frame for `IPC::CreateMutex` as being different from the frame for the kill call. At least, that is my current conjecture. I have noticed the stack frame doing a syscall through glibc absent from stack traces when syscalls were recorded during perf profiling in the past, but until now, I have never looked into the possible cause.

**Comment by ryao** [comment]
If Valve cannot fix this any time soon, would Valve please open source vrclient.so so that I could fix it for them?

**Comment by ryao** [comment]
Alternatively, I would be happy to settle for a “If you are such a big fan of our software that you would go through the hassle of reverse engineering and modifying binary code to fix this before we do, we will not sue you”.

**Comment by ryao** [comment]
That does not work with the Nvidia driver and I am already using perf events manually. It might be useful for non-VR things on my systems that use Intel graphics though. Thanks.

**Comment by Plagman** [comment]
gpuvis is a frontend to visualize any perf events, with specialized visualizers for CPU and GPU scheduling events, and print events. You're right that there's no NVIDIA GPU scheduling events to visualize at the moment, but it's still valuable on the CPU/print side even on NVIDIA machines.

**Comment by lostgoat** [comment]
This issue should be fixed in SteamVR 1.7.2

**Comment by ryao** [comment]
I can confirm that it is fixed. :)
